x11rb-protocol/src/parse_display/mod.rs

```diff
@@ -77,6 +84,41 @@ fn parse_display_impl(dpy_name: &str) -> Option<ParsedDisplay> {
     })
 }
 
+// Check for "launchd mode" where we get the full path to a unix socket
+fn parse_display_direct_path(dpy_name: &str) -> Option<ParsedDisplay> {
+    #[cfg(feature = "std")]
+    fn file_exists(path: impl AsRef<std::path::Path>) -> bool {
+        path.as_ref().exists()
+    }
+    #[cfg(not(feature = "std"))]
+    fn file_exists(_path: &str) -> bool {
+        // Just hope for the best
+        true
+    }
```

Comment:
I would think that this could be moved to `x11rb` itself. Just return both of the possible paths and let the end consumer decide whether or not they exist.

Reply:
How would I return both possible paths? The current API can only return a single result and I have no good idea how to change that without ending up with an ugly result.

We could prefer parsing `/tmp/foo:42` as `/tmp/foo` with screen `42`, but that would be a libxcb incompatibility since it prefers the socket with name `/tmp/foo:42` and screen 0 in this case.

But I guess at least in the `no_std` case, it makes more sense to parse a screen number if one is present. At least I feel like that case is more likely than putting a colon into the path. But for the `std` case, I would prefer to stay compatible with libxcb.

Reply:
Another idea would be to have two APIs:

- The existing API is only available when `std` is enabled.
- A second function is always available and needs a `check_file_exists: impl Fn(&str) -> bool` as another argument

Reply:
> Another idea would be to have two APIs:
> 
>     * The existing API is only available when `std` is enabled.
> 
>     * A second function is always available and needs a `check_file_exists: impl Fn(&str) -> bool` as another argument

I like that. Ideally, parsing should be a "pure function" (i.e., output only depends on input). But considering that it looks like display cannot be unambiguously parsed without context, at least it allows the caller to control the context.

Reply:
What if the API instead returns an `Iterator` over the `ParsedDisplay`s? That way it could return only one display for normal cases, but two to check for this case? That way the file existence check can be deferred until later, in `x11rb`.

Reply:
I updated the code with the extra callback function.

Sorry, but I don't really like the `Iterator` idea. It makes the API more complicated in the common case (just use `std` to check for file existence). With this new version, the common case (`std` is enabled) is just easy and the `no_std` case is not too complicated either.